pym/bob/layers.py

```diff
@@ -0,0 +1,228 @@
+        config = self.loadYaml(configYaml, (RecipeSet.STATIC_CONFIG_SCHEMA, b''))
+        for l in config.get('layers', []):
+            if not isinstance(l, dict): continue
+            scmSpec = l.get("scmSpec", {})
+            scmSpec.update({'recipe':configYaml})
+            layerScms = Scm(scmSpec,
+                            self.__recipes.getRootEnv(),
+                            overrides=self.__recipes.scmOverrides(),
+                            recipeSet=self.__recipes)
+            self.__subLayers.append(Layer(l.get("name"),
+                                          self.__root,
+                                          self.__recipes,
+                                          self.__yamlCache,
+                                          layerScms))
+
+    def getSubLayers(self):
+        return self.__subLayers
+
+    def status(self, verbose):
+        if self.__scm is None:
+            return
+        pp = PackagePrinter(verbose, False, False)
+        status = self.__scm.status(self.__layerDir)
+        pp.show(status, self.__layerDir)
+
+class Layers:
+    def __init__(self, recipes, loop, yamlCache=None):
+        self.__layers = {}
+        self.__loop = loop
+        self.__scmUpdate = False
+        self.__recipes = recipes
+        self.__yamlCache = yamlCache
+        if yamlCache is None:
+            self.__yamlCache = YamlCache()
+
+    def __haveLayer(self, layer):
+        for depth,layers in self.__layers.items():
+            for l in layers:
+                if l.getName() == layer.getName():
+                    return True
+        return False
+
+    def __collect(self, depth, update, verbose):
+        self.__layers[depth+1] = []
+        newLevel = False
+        for l in self.__layers[depth]:
+            if update:
+                l.checkout(self.__loop, verbose)
+            l.parse()
+            for subLayer in l.getSubLayers():
+                if not self.__haveLayer(subLayer):
+                    self.__layers[depth+1].append(subLayer)
+                    newLevel = True
+        if newLevel:
+            self.__collect(depth + 1, update, verbose)
+
+    def collect(self, update, verbose=0):
+        if self.__yamlCache is not None:
+            self.__yamlCache.open()
+        try:
+            rootLayers = Layer("", os.getcwd(), self.__recipes, self.__yamlCache)
+            rootLayers.parse()
+            self.__layers[0] = rootLayers.getSubLayers();
+            self.__collect(0, update, verbose)
+        finally:
+            if self.__yamlCache is not None:
+                self.__yamlCache.close()
+
+    def status(self, verbose):
+        for level in self.__layers:
+            for layer in self.__layers[level]:
+                layer.status(verbose)
+
+def fetchLayers(recipes, yamlCache, loop, verbose):
+    layers = Layers(recipes, loop, yamlCache)
+    layers.collect(True, verbose)
+
+def doLayers(argv, bobRoot):
```

Comment:
The layers command handling should probably be moved into `pym/bob/cmds/layers.py` next to the other commands.